File: ui-tests/PageObjects/Panes/CallTrace/CallTraceArgument.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Microsoft.Playwright;
using UiTests.PageObjects.Components;
using UiTests.Utils;

namespace UiTests.PageObjects.Panes.CallTrace;

/// <summary>
/// Represents a single argument entry within a call trace record.
/// </summary>
public class CallTraceArgument
{
    private static readonly string[] DefaultOptions = { "Add value to scratchpad" };

    private readonly CallTracePane _pane;
    private readonly ILocator _root;
    private readonly ContextMenu _contextMenu;

    public CallTraceArgument(CallTracePane pane, ILocator root, ContextMenu contextMenu)
    {
        _pane = pane ?? throw new ArgumentNullException(nameof(pane));
        _root = root ?? throw new ArgumentNullException(nameof(root));
        _contextMenu = contextMenu ?? throw new ArgumentNullException(nameof(contextMenu));
    }

    /// <summary>
    /// Locator representing the argument container.
    /// </summary>
    public ILocator Root => _root;

    private ILocator NameLocator() => _root.Locator(".call-arg-name");
    private ILocator ValueLocator() => _root.Locator(".call-arg-text");

    /// <summary>
    /// Argument name as displayed in the UI.
    /// </summary>
    public async Task<string> NameAsync()
    {
        var text = await NameLocator().InnerTextAsync();
        return text?.TrimEnd('=').Trim() ?? string.Empty;
    }

    /// <summary>
    /// Value text rendered for the argument.
    /// </summary>
    public async Task<string> ValueAsync()
    {
        var text = await ValueLocator().InnerTextAsync();
        return text?.Trim() ?? string.Empty;
    }

    /// <summary>
    /// Opens the context menu attached to the argument.
    /// </summary>
    public async Task<ContextMenu> OpenContextMenuAsync()
    {
        await _root.ClickAsync(new() { Button = MouseButton.Right });
        await _contextMenu.WaitForVisibleAsync();
        return _contextMenu;
    }

    /// <summary>
    /// Reads the context menu entries for the argument.
    /// </summary>
    public async Task<IReadOnlyList<string>> ContextMenuEntriesAsync()
    {
        var menu = await OpenContextMenuAsync();
        var entries = await menu.GetEntriesAsync();
        await menu.DismissAsync();
        return entries.Select(e => e.Text).ToList();
    }

    /// <summary>
    /// Expected argument context menu entries.
    /// </summary>
    public IReadOnlyList<string> ExpectedContextMenuEntries => DefaultOptions;

    /// <summary>
    /// Adds the argument value to the scratchpad using the context menu.
    /// </summary>
    public async Task AddToScratchpadAsync()
    {
        var menu = await OpenContextMenuAsync();
        await menu.SelectAsync(DefaultOptions[0]);
    }

    /// <summary>
    /// Opens the inline tooltip rendering the expanded value.
    /// </summary>
    public async Task<ValueComponentView?> OpenTooltipAsync()
    {
        await _root.ClickAsync();
        await RetryHelpers.RetryAsync(async () => (await _pane.ActiveTooltipAsync()) is not null);
        return await _pane.ActiveTooltipAsync();
    }
}

File: tsc-ui-tests/page-objects/panes/call-trace/call-trace-argument.ts
import type { Locator } from "@playwright/test";
import { ContextMenu } from "../../components/context-menu";
import type { ValueComponentView } from "../../components/value-component-view";
import type { CallTracePane } from "./call-trace-pane";
import { retry } from "../../../lib/retry-helpers";

const DEFAULT_CONTEXT_MENU_ENTRIES = ["Add value to scratchpad"];

/**
 * Represents a single argument entry within a call trace record.
 *
 * Port of ui-tests/PageObjects/Panes/CallTrace/CallTraceArgument.cs
 */
export class CallTraceArgument {
  private readonly pane: CallTracePane;
  readonly root: Locator;
  private readonly contextMenu: ContextMenu;

  constructor(pane: CallTracePane, root: Locator, contextMenu: ContextMenu) {
    this.pane = pane;
    this.root = root;
    this.contextMenu = contextMenu;
  }

  private nameLocator(): Locator {
    return this.root.locator(".call-arg-name");
  }

  private valueLocator(): Locator {
    return this.root.locator(".call-arg-text");
  }

  async name(): Promise<string> {
    const text = await this.nameLocator().innerText();
    return text.replace(/=$/, "").trim();
  }

  async value(): Promise<string> {
    return (await this.valueLocator().innerText()).trim();
  }

  async openContextMenu(): Promise<ContextMenu> {
    await this.root.click({ button: "right" });
    await this.contextMenu.waitForVisible();
    return this.contextMenu;
  }

  async contextMenuEntries(): Promise<string[]> {
    const menu = await this.openContextMenu();
    const entries = await menu.getEntries();
    await menu.dismiss();
    return entries.map((e) => e.text);
  }

  get expectedContextMenuEntries(): readonly string[] {
    return DEFAULT_CONTEXT_MENU_ENTRIES;
  }

  async addToScratchpad(): Promise<void> {
    const menu = await this.openContextMenu();
    await menu.select(DEFAULT_CONTEXT_MENU_ENTRIES[0]);
  }

  async openTooltip(): Promise<ValueComponentView | null> {
    await this.root.click();
    await retry(async () => (await this.pane.activeTooltip()) !== null);
    return this.pane.activeTooltip();
  }
}
